src/ir-util.h

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright 2016 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef WABT_IR_UTIL_H_
+#define WABT_IR_UTIL_H_
+
+#include "src/common.h"
+#include "src/ir.h"
+
+namespace wabt {
+
+struct Label {
+  Label(LabelType label_type,
+        const std::string& name,
+        const TypeVector& param_types,
+        const TypeVector& result_types)
+      : name(name),
+        label_type(label_type),
+        param_types(param_types),
+        result_types(result_types) {}
+
+  std::string name;
+  LabelType label_type;
+  TypeVector param_types;
+  TypeVector result_types;
+};
+
+struct ModuleContext {
```

Comment:
A base class name doesn't necessarily need to express the smallest possible set of subclasses, it can be quite broad. Is an Animal an IntrusiveLinkedListNode? It sure can be.

This is a ModuleContext in the sense that it tracks additional information as you iterate thru a Module, but it is not necessarily specific to text writing.

Also, goal was to make as little changes to WatWriter by making it a base class, not a member.